### PR TITLE
feat: support X-APPLE-CALENDAR-COLOR

### LINKED ISSUE #143: [feature] Support color in hex representation (X-APPLE-CALENDAR-COLOR)
Support for `X-APPLE-CALENDAR-COLOR`would be useful to allow hex-codes in addition to just CSS color names (as introduced in #30).

Also see: bitfireAT/icsx5#31

### PR DESCRIPTION
**Pull request by arran4**
Closes #143 by adding support for `X-APPLE-CALENDAR-COLOR` alongside `COLOR`.

---
*PR created automatically by Jules for task [4842174814514569679](https://jules.google.com/task/4842174814514569679) started by @arran4*

### CODE REVIEW

File: calendar.go [hunk 1]
```diff
@@ -947,7 +953,7 @@ func ParseCalendarWithOptions(r io.Reader, options ...any) (*Calendar, error) {
 				}
 			case "BEGIN":
 				state = "components"
-			case string(PropertyCalscale), string(PropertyMethod), string(PropertyProductId), string(PropertyVersion), string(PropertyName), string(PropertyXWRCalName), string(PropertyXWRCalDesc), string(PropertyXWRTimezone), string(PropertyXWRCalID), string(PropertyXPublishedTTL), string(PropertyRefreshInterval), string(PropertyColor), string(PropertyDescription), string(PropertyLastModified), string(PropertyUrl), string(PropertyTzid), string(PropertyTimezoneId), string(PropertySource):
+			case string(PropertyCalscale), string(PropertyMethod), string(PropertyProductId), string(PropertyVersion), string(PropertyName), string(PropertyXWRCalName), string(PropertyXWRCalDesc), string(PropertyXWRTimezone), string(PropertyXWRCalID), string(PropertyXPublishedTTL), string(PropertyRefreshInterval), string(PropertyColor), string(PropertyXAppleCalendarColor), string(PropertyDescription), string(PropertyLastModified), string(PropertyUrl), string(PropertyTzid), string(PropertyTimezoneId), string(PropertySource):
```
[the remark this answers is not in the document]
Comment (arran4) [reply calendar.go]: @jules It's an X property (experimental) so it should be handled by the default.  So this line doesn't need to be changed.

File: calendar.go [hunk 2]
```diff
@@ -582,6 +583,11 @@ func (cal *Calendar) SetName(s string, params ...PropertyParameter) {
 
 func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
 	cal.setProperty(PropertyColor, s, params...)
+	cal.setProperty(PropertyXAppleCalendarColor, s, params...)
+}
+
+func (cal *Calendar) SetXAppleCalendarColor(s string, params ...PropertyParameter) {
```
Comment (arran4) [inline calendar.go]: @jules can we get some color.Color based helpers?

File: calendar.go [hunk 3]
```diff
@@ -582,6 +583,11 @@ func (cal *Calendar) SetName(s string, params ...PropertyParameter) {
 
 func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
 	cal.setProperty(PropertyColor, s, params...)
+	cal.setProperty(PropertyXAppleCalendarColor, s, params...)
```
Comment (arran4) [inline calendar.go]: @jules we shouldn't be setting two properties on a single simplely named setter especially a normal + experimental.

**Comment by arran4** [comment]
@jules please add the appropriate getters, and run go fmt.

Can you also think about other color formats which might be required such as a to/from hex (which might be just a wrapper function on color or a type switched colour input.)

File: calendar.go [hunk 4]
```diff
@@ -584,6 +586,23 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
...
+}
+
+func colorToHex(c color.Color) string {
+	r, g, b, _ := c.RGBA()
+	return fmt.Sprintf("#%02x%02x%02x", uint8(r>>8), uint8(g>>8), uint8(b>>8))
+}
+
+func (cal *Calendar) SetColorFromColor(c color.Color, params ...PropertyParameter) {
+	cal.setProperty(PropertyColor, colorToHex(c), params...)
```
Comment (arran4) [inline calendar.go]: @jules typeswitched and please consider where we include alpha too and start with a `#`

File: calendar.go [hunk 5]
```diff
@@ -653,6 +687,15 @@ func (cal *Calendar) addComponent(c Component) {
 	cal.Components = append(cal.Components, c)
 }
 
+func (cal *Calendar) GetProperty(property Property) *CalendarProperty {
```
Comment (arran4) [inline calendar.go:692]: @jules I was expecting getter variants of what we have added.

File: calendar.go [hunk 6]
```diff
@@ -653,6 +687,23 @@ func (cal *Calendar) addComponent(c Component) {
...
+	for i := range cal.CalendarProperties {
+		if cal.CalendarProperties[i].IANAToken == string(property) {
+			return &cal.CalendarProperties[i]
+		}
+	}
+	return nil
+}
+
+func (cal *Calendar) GetColor() *CalendarProperty {
```
Comment (arran4) [inline calendar.go:701]: @jules I am expecting the return type to be useful to the user. perhaps variants: 
* `GetColorAsRBGString`
* `...AsColor`
* etc

File: calendar.go [hunk 7]
```diff
@@ -653,6 +687,23 @@ func (cal *Calendar) addComponent(c Component) {
...
+	return cal.GetProperty(PropertyColor)
+}
+
+func (cal *Calendar) GetXAppleCalendarColor() *CalendarProperty {
```
Comment (arran4) [inline calendar.go]: @jules I am expecting the return type to be useful to the user. perhaps variants: 
* `GetColorAsRBGString`
* `...AsColor`
* etc

File: calendar.go [hunk 8]
```diff
@@ -584,6 +586,63 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
...
+		return fmt.Sprintf("#%02x%02x%02x", c.R, c.G, c.B)
+	}
+	r, g, b, a := c.RGBA()
+	if a < 65535 {
+		return fmt.Sprintf("#%02x%02x%02x%02x", uint8(r>>8), uint8(g>>8), uint8(b>>8), uint8(a>>8))
+	}
+	return fmt.Sprintf("#%02x%02x%02x", uint8(r>>8), uint8(g>>8), uint8(b>>8))
+}
+
+// hexToColor takes a hex color string like "#FF0000" or "#FF0000FF" and returns a color.Color.
```
Comment (arran4) [inline calendar.go]: @jules are there any go sdk functions we could use instead? If not this should be fully tested.

File: calendar.go [hunk 9]
```diff
@@ -584,6 +586,63 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
 	cal.setProperty(PropertyColor, s, params...)
 }
 
+func (cal *Calendar) SetXAppleCalendarColor(s string, params ...PropertyParameter) {
+	cal.setProperty(PropertyXAppleCalendarColor, s, params...)
+}
+
+func colorToHex(c color.Color) string {
```
Comment (arran4) [inline calendar.go]: @jules are there any go sdk functions we could use instead? If not this should be fully tested.

**Comment by arran4** [comment]
Need review by devs on apple products.

**Comment by arran4** [comment]
@jules please import the tests: https://github.com/bitfireAT/icsx5/pull/32/changes

**Comment by arran4** [comment]
@jules add tests for X11 colours for the standard COLOR prop

File: calendar.go [hunk 10]
```diff
@@ -653,6 +712,39 @@ func (cal *Calendar) addComponent(c Component) {
...
+func (cal *Calendar) GetColor() *CalendarProperty {
+	return cal.GetProperty(PropertyColor)
+}
+
+func (cal *Calendar) GetColorAsColor() (color.Color, error) {
+	p := cal.GetColor()
+	if p == nil {
+		return nil, errors.New("color property not found")
+	}
+	return hexToColor(p.Value)
```
Comment (arran4) [inline calendar.go]: Color using X11 naming, so this can only be AsString @jules

File: calendar.go [hunk 11]
```diff
@@ -653,6 +712,39 @@ func (cal *Calendar) addComponent(c Component) {
...
+}
+
+func (cal *Calendar) GetXAppleCalendarColor() *CalendarProperty {
```
Comment (arran4) [inline calendar.go]: @jules the default format is as a string which is a hex code, so we should add a helper that specifies "as hex string"

File: calendar.go [hunk 12]
```diff
@@ -584,6 +586,63 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
 	cal.setProperty(PropertyColor, s, params...)
 }
 
+func (cal *Calendar) SetXAppleCalendarColor(s string, params ...PropertyParameter) {
```
Comment (arran4) [inline calendar.go]: @jules we need AsString for the normal COLOR prop

File: calendar.go [hunk 13]
```diff
@@ -584,6 +586,63 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
...
+	}
+	r, g, b, a := c.RGBA()
+	if a < 65535 {
+		return fmt.Sprintf("#%02x%02x%02x%02x", uint8(r>>8), uint8(g>>8), uint8(b>>8), uint8(a>>8))
+	}
+	return fmt.Sprintf("#%02x%02x%02x", uint8(r>>8), uint8(g>>8), uint8(b>>8))
+}
+
+// hexToColor takes a hex color string like "#FF0000" or "#FF0000FF" and returns a color.Color.
+func hexToColor(s string) (color.Color, error) {
```
Comment (arran4) [inline calendar.go]: @jules color specific non-calendar integrated functions, consts, tests, etc should be in colors.go and colors_test.go

File: calendar.go [hunk 14]
```diff
@@ -653,6 +667,47 @@ func (cal *Calendar) addComponent(c Component) {
...
+		}
+	}
+	return nil
+}
+
+func (cal *Calendar) GetColor() *CalendarProperty {
+	return cal.GetProperty(PropertyColor)
+}
+
+func (cal *Calendar) GetColorAsString() string {
```
Comment (arran4) [inline calendar.go:705]: @jules `AsColorName() (ColorName, ok)` and `AsNearestColorName()`

File: color.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
+package ics
+
+import (
+	"fmt"
+	"image/color"
+)
+
+func colorToHex(c color.Color) string {
```
Comment (arran4) [inline color.go]: @jules create a `type ColorName string`  and `type W3CColorName string` 

It should have a mapping for all X11 https://en.wikipedia.org/wiki/X11_color_names  such that there are functions (`ColorName.To...()` `ColorNameFrom...()` that
*  Closest color.Color or HexString lookup to ColorName 
* Exact color.Color or HexString lookup to ColorName 
* ColorName to color.Color or HexString lookup

All tested. Please also include all alternative names. In the case of w3c / x11 clash, append X11 to the end of the X11 name

File: color.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+func ClosestColorNameFromHexString(hex string) (ColorName, error) {
+	c, err := hexToColor(hex)
+	if err != nil {
+		return "", err
+	}
+	return ClosestColorNameFromColor(c), nil
+}
+
+// ClosestColorNameFromColor returns the closest ColorName for a given color.Color using Euclidean distance in RGB space.
+func ClosestColorNameFromColor(c color.Color) ColorName {
```
Comment (arran4) [inline color.go:315]: @jules is this the best algorithm we could use?

File: color.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+		return color.RGBA{R: r, G: g, B: b, A: a}, nil
+	default:
+		return nil, fmt.Errorf("invalid hex color length: %s", s)
+	}
+}
+
+// Standard mapping of X11/CSS3 color names to hex codes.
+// Note: where W3C/CSS and X11 clash (e.g. green, gray), we default to the W3C spec as it's standard for web,
+// but include the alternate names.
+var colorNameToHex = map[ColorName]string{
```
Comment (arran4) [inline color.go]: @jules we need another table for aliases right?

File: color.go [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+	"strings"
+)
+
+// ColorName represents a CSS/X11 standard named color string.
+type ColorName string
```
Comment (arran4) [inline color.go:11]: @jules `type ColorHex string` too

File: color_test.go
Comment (arran4) [inline color_test.go:1]: @jules I am expecting 100% test coverage on colors.

File: calendar.go [hunk 19]
```diff
@@ -584,6 +586,18 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
 	cal.setProperty(PropertyColor, s, params...)
 }
 
+func (cal *Calendar) SetXAppleCalendarColor(s string, params ...PropertyParameter) {
+	cal.setProperty(PropertyXAppleCalendarColor, s, params...)
+}
+
+func (cal *Calendar) SetColorFromColor(c color.Color, params ...PropertyParameter) {
+	cal.setProperty(PropertyColor, string(colorToHex(c)), params...)
```
Comment (arran4) [inline calendar.go]: @jules color uses ColorName not ColorHex

File: calendar.go [hunk 20]
```diff
@@ -584,6 +586,18 @@ func (cal *Calendar) SetColor(s string, params ...PropertyParameter) {
...
+}
+
+func (cal *Calendar) SetXAppleCalendarColorFromColor(c color.Color, params ...PropertyParameter) {
```
Comment (arran4) [inline calendar.go]: @jules need a ColorName variant

**Comment by arran4** [comment]
@jules clean up coverage file

File: color.go [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+	r1, g1, b1 := rgbaToFloat(c1)
+	r2, g2, b2 := rgbaToFloat(c2)
+	dr := r1 - r2
+	dg := g1 - g2
+	db := b1 - b2
+	return math.Sqrt(dr*dr + dg*dg + db*db)
+}
+
+// ClosestColorNameFromHexString returns the closest ColorName for a given hex string using Euclidean distance in RGB space.
+func ClosestColorNameFromHexString(hex ColorHex) (ColorName, error) {
```
Comment (arran4) [inline color.go:306]: @jules ensure coverage of this funciton

File: color.go [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+	c, err := hexToColor(string(hex))
+	if err != nil {
+		return "", err
+	}
+	return ClosestColorNameFromColor(c), nil
+}
+
+// ClosestColorNameFromColor returns the closest ColorName for a given color.Color using Euclidean distance in RGB space.
+func ClosestColorNameFromColor(c color.Color) ColorName {
```
Comment (arran4) [inline color.go:315]: @jules ensure coverage of this funciton

File: color_test.go [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+				return
+			}
+			if !tt.wantErr && got != tt.want {
+				t.Errorf("hexToColor() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestColorName(t *testing.T) {
```
Comment (arran4) [inline color_test.go:63]: @jules these tests should be their own func, and we should use more go ideomatic testing so we can test more edge cases etc. Using t.Run outside of a loop without a good use case makes the code hard to follow and nonideomatic go.
